Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs

```diff
@@ -0,0 +1,215 @@
+﻿// Copyright 2019 Google LLC
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     https://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Cloud.Storage.V1;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.AspNetCore.DataProtection.Repositories;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Xml;
+using System.Xml.Linq;
+
+using Object = Google.Apis.Storage.v1.Data.Object;
+
+namespace Google.Cloud.AspNetCore.DataProtection.Storage
+{
+    /// <summary>
+    /// Implementation of <see cref="IXmlRepository"/> that stores the protected elements in a single Google Cloud Storage file.
+    /// This class is configured by <see cref="GoogleCloudDataProtectionBuilderExtensions.PersistKeysToGoogleCloudStorage(IDataProtectionBuilder, StorageClient, string, string)" />
+    /// (and other overloads).
+    /// </summary>
+    internal sealed partial class CloudStorageXmlRepository : IXmlRepository
+    {
+        /// <summary>
+        /// Retry handler for loading the latest document.
+        /// </summary>
+        private static readonly RetryHandler s_loadLatestRetryHandler = new RetryHandler(
+            maxAttempts: 5,
+            initialBackoff: TimeSpan.FromSeconds(0.2),
+            backoffMultiplier: 1.5,
+            maxBackoff: TimeSpan.FromSeconds(1));
+
+        /// <summary>
+        /// Retry handler for storing changes. This has longer backoffs than the "load latest" retry handler,
+        /// as object updates are rate-limited.
+        /// </summary>
+        private static readonly RetryHandler s_storeRetryHandler = new RetryHandler(
+            maxAttempts: 5,
+            initialBackoff: TimeSpan.FromSeconds(1),
+            backoffMultiplier: 2.0,
+            maxBackoff: TimeSpan.FromSeconds(3));
+
+        private readonly StorageClient _client;
+        private readonly string _bucketName;
+        private readonly string _objectName;
+
+        private readonly object _lock = new object();
+        private LoadDocumentResult _latestDocument;
```

Comment:
Keeping _latestDocument adds a lot of complexity to the code, and I'm not sure of the benefit.
In what sequence of calls will caching the the latest document yield a benefit?

Reply:
Anything that calls GetElements many times, which I think may happen. It still makes a metadata request per call, but doesn't fetch the data.